### PR TITLE
Add QUIC stateless reset test

### PR DESCRIPTION
**Pull request by nhorman**
QUIC supports the concept of stateless reset, in which a specially
crafted frame is sent to a client informing it that the QUIC state
information is no longer available, and the connection should be closed
immediately.  Test for proper client support here

##### Checklist
- [x] tests are added or updated

### CODE REVIEW

File: test/quic_multistream_test.c
Comment (hlandau) [inline test/quic_multistream_test.c:5125]: nit: extra line

File: test/quic_multistream_test.c
Comment (hlandau) [inline test/quic_multistream_test.c:5135]: nit: extra line

File: test/quic_multistream_test.c
Comment (hlandau) [inline test/quic_multistream_test.c:5181]: drop fprintfs

File: test/quic_multistream_test.c
Comment (hlandau) [inline test/quic_multistream_test.c:5173]: drop fprintfs or use reporting macros in testutil.h

File: test/quic_multistream_test.c
Comment (hlandau) [inline test/quic_multistream_test.c:5202]: The rationale behind doing a write and read-expect is to make sure an exchange of packets has occurred and insure that some thing we are trying to inject with the fault injector has a chance to be sent, pass through an OS socket, and be received. The OS socket can have somewhat variable behaviour even over the loopback interface, especially on macOS, so this avoids stochastic test failures on macOS.

Therefore, I think this is the wrong way around; since you are sending a NCID frame to the client and need to make sure the client has had a chance to receive it, I would do the write on the server side and the read_expect (which will wait around as needed) on the client side.

nit: extra space after comma

Comment (hlandau) [inline test/quic_multistream_test.c:5202]: The rationale behind doing a write and read-expect is to make sure an exchange of packets has occurred and insure that some thing we are trying to inject with the fault injector has a chance to be sent, pass through an OS socket, and be received. The OS socket can have somewhat variable behaviour even over the loopback interface, especially on macOS, so this avoids stochastic test failures on macOS.

Therefore, I think this is the wrong way around; since you are sending a NCID frame to the client and need to make sure the client has had a chance to receive it, I would do the write on the server side and the read_expect (which will wait around as needed) on the client side.

**Review by hlandau — COMMENTED** [review]
Great work! Just minor nits and one substantive comment and I'll approve.

You'll want to take a look at backporting this test to 3.2 IMO.

File: test/quic_multistream_test.c
Comment (nhorman) [reply test/quic_multistream_test.c]: fixed

File: test/quic_multistream_test.c
Comment (nhorman) [reply test/quic_multistream_test.c:5134]: fixed

File: test/quic_multistream_test.c
Comment (nhorman) [reply test/quic_multistream_test.c]: dropped

File: test/quic_multistream_test.c
Comment (nhorman) [reply test/quic_multistream_test.c]: dropped

File: test/quic_multistream_test.c
Comment (nhorman) [reply test/quic_multistream_test.c]: had to swap a few ops around to avoid a failure in looking up the stream id, but its doing sends from the server now

File: test/quic_multistream_test.c
Comment (hlandau) [inline test/quic_multistream_test.c:5201]: Surely this should be before the OP_S_WRITE?

**Review by hlandau — CHANGES_REQUESTED** [review]

**Comment by nhorman** [comment]
hmm, it would seem the macos tests are timing out waiting for the connection close condition to be detected.  Unfortunately I cant OP_[S|C]_WRITE anything to force a socket read, as that fails prior to the expecation of the close detection.  OP_{C|S]_READ_FAIL looks like it might be promising, but those also timeout detecting a close condition.  Thoughts?

**Comment by nhorman** [comment]
@hlandau, scratch that, found a way around it.  The close condition on the client is consistent and suffices to confirm.thw reset.

Side note: there is a thread race that occurs when the inject function is changed during a test, but I chose to avoid it by setting the function at the start of the script, rather than introduce a bunch of new locking

File: test/quic_multistream_test.c
Comment (hlandau) [inline test/quic_multistream_test.c:5151]: Replace `h->s_priv` with `ACQUIRE_S()` here

**Review by hlandau — CHANGES_REQUESTED** [review]
One last thing

**Comment by nhorman** [comment]
ACQUIRE_S() doesn't work in injector functions as helper_local isn't passed. Will need to find a way around that

**Review by hlandau — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by nhorman** [comment]
Note: tested locally against 3.2 here and the test fails.  It appears to be occuring because, during stateless reset injection on 3.2 demux_process_pending_urxe issues a callback to reset_token_cb to determine if the frame is an SR, and, if so, returns -1, whcih goes back up the call stack, causing an error in SSL_inject_net_dgram, resulting in a failed test.  The master branch silently discards SR frames in port_default_packet_handler after closing the connection as a side effect.  Point being, backporting to 3.2 will take some additional work in the quic code to operate properly

**Comment by mattcaswell** [comment]
Although this is "ready-to-merge" - it seems its not really. At least not to 3.2 since tests would fail if it is merged there. I guess we're blocked on having fixes for the 3.2 issues.

**Comment by hlandau** [comment]
You can merge it to master. 3.2 is a separate issue and PR.

**Comment by mattcaswell** [comment]
I've removed the 3.2 label for now to prevent an accidental merge there (although I believe the approvals above indicate it can be merged there once 3.2 is fixed)

**Comment by nhorman** [comment]
merged

**Comment by hlandau** [comment]
@nhorman IMO create a separate issue for the 3.2 investigation and close this.

**Comment by nhorman** [comment]
Ack will do

**Pull request closed without merging**
